File: scripts/lib/ai-fiction.ts
/**
 * AI Fiction Generation for Daily Reflections
 *
 * Generates a short, engaging fiction piece for the daily reflection note.
 * The fiction is thematically inspired by the day's content — books, videos,
 * blog posts, etc. — abstracted into creative themes.
 *
 * Style rules:
 * - Each sentence begins with an emoji
 * - No quotation marks
 * - Under 100 words
 * - Evocative, slightly philosophical tone
 *
 * Pure functions handle content prep and application; Gemini handles creative
 * fiction writing.
 *
 * @module ai-fiction
 */

import { isRetriableError } from "../generate-blog-post.ts";
import {
  TWEET_SECTION_HEADER,
  BLUESKY_SECTION_HEADER,
  MASTODON_SECTION_HEADER,
} from "./types.ts";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

export const FICTION_SECTION_HEADER = "## 🤖🐲 AI Fiction";
export const DEFAULT_FICTION_MODEL = "gemini-2.5-flash";

const GEMINI_MAX_RETRIES = 3;
const GEMINI_BASE_DELAY_MS = 2_000;

const EMBED_HEADERS = [
  TWEET_SECTION_HEADER,
  BLUESKY_SECTION_HEADER,
  MASTODON_SECTION_HEADER,
] as const;

// ---------------------------------------------------------------------------
// Pure Functions
// ---------------------------------------------------------------------------

/**
 * Strip frontmatter and embed sections from reflection content,
 * leaving only the substantive content for fiction prompt construction.
 */
export const stripForPrompt = (content: string): string => {
  const lines = content.split("\n");

  // Strip frontmatter
  const hasFrontmatter = lines[0]?.trim() === "---";
  const fmEnd = hasFrontmatter
    ? lines.findIndex((l, i) => i > 0 && l.trim() === "---")
    : -1;
  const bodyLines = fmEnd >= 0 ? lines.slice(fmEnd + 1) : lines;
  const body = bodyLines.join("\n");

  // Strip embed sections (everything from first embed header onward)
  const firstEmbedIndex = EMBED_HEADERS
    .map((header) => body.indexOf(header))
    .filter((index) => index >= 0)
    .reduce((min, index) => Math.min(min, index), body.length);

  // Also strip the existing AI Fiction section if present
  const fictionIndex = body.indexOf(FICTION_SECTION_HEADER);
  const cutoff = fictionIndex >= 0
    ? Math.min(fictionIndex, firstEmbedIndex)
    : firstEmbedIndex;

  // Strip the Updates section if present
  const updatesIndex = body.indexOf("## 🔄 Updates");
  const finalCutoff = updatesIndex >= 0
    ? Math.min(updatesIndex, cutoff)
    : cutoff;

  return body.slice(0, finalCutoff).trim();
};

/**
 * Check whether a reflection note already has an AI Fiction section.
 */
export const reflectionNeedsFiction = (content: string): boolean =>
  !content.includes(FICTION_SECTION_HEADER);

/**
 * Build the Gemini prompt for fiction generation.
 */
export const buildFictionPrompt = (
  strippedContent: string,
): { readonly system: string; readonly user: string } => {
  const system = `You are a creative fiction writer. Your task is to:

1. Read the content from today's daily reflection note
2. Identify 2-4 abstract themes from the content
3. Write a short piece of engaging fiction (UNDER 100 words) inspired by those themes

RULES:
- Every sentence MUST begin with an emoji
- NEVER use quotation marks (no ", no ', no \`, no curly quotes)
- Keep it under 100 words
- The fiction should be evocative and slightly philosophical
- Abstract the themes — do NOT directly reference the specific books, videos, or articles
- Do NOT include any heading or title — just the fiction text
- Output ONLY the fiction — no explanation, no theme list, no preamble`;

  const user = `Write a short fiction piece inspired by themes from this daily reflection:\n\n${strippedContent}`;

  return { system, user };
};

/**
 * Parse and clean the raw Gemini response.
 * Strips code fences, removes quotation marks, and trims whitespace.
 */
export const parseFictionResponse = (raw: string): string =>
  raw
    .replace(/^```(?:markdown|md)?\s*\n/, "")
    .replace(/\n```\s*$/, "")
    .replace(/^## 🤖🐲 AI Fiction\s*\n*/m, "")
    .replace(/[\u0022\u0027\u0060\u2018\u2019\u201C\u201D]/g, "")
    .trim();

/**
 * Insert an AI Fiction section into the reflection note content.
 * Places it just before the first embed section, or at the end.
 */
export const applyFiction = (content: string, fiction: string): string => {
  const sectionBlock = `${FICTION_SECTION_HEADER}\n\n${fiction}`;

  // Find the first embed section to insert before
  const embedIndices = EMBED_HEADERS
    .map((header) => content.indexOf(header))
    .filter((index) => index >= 0);
  const firstEmbedIndex = embedIndices.length > 0 ? Math.min(...embedIndices) : -1;

  // Also check for Updates section — insert before it too
  const updatesIndex = content.indexOf("## 🔄 Updates");
  const candidates = [firstEmbedIndex, updatesIndex].filter((i) => i >= 0);
  const insertBeforeIndex = candidates.length > 0 ? Math.min(...candidates) : -1;

  return insertBeforeIndex >= 0
    ? `${content.slice(0, insertBeforeIndex).trimEnd()}\n\n${sectionBlock}\n\n${content.slice(insertBeforeIndex)}`
    : `${content.trimEnd()}\n\n${sectionBlock}\n`;
};

// ---------------------------------------------------------------------------
// Gemini API
// ---------------------------------------------------------------------------

const delay = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms));

const callGeminiOnce = async (
  apiKey: string,
  model: string,
  prompt: { readonly system: string; readonly user: string },
): Promise<string> => {
  const { GoogleGenAI } = await import("@google/genai");
  const ai = new GoogleGenAI({ apiKey });
  const contents = [
    { role: "user" as const, parts: [{ text: `${prompt.system}\n\n${prompt.user}` }] },
  ];
  const result = await ai.models.generateContent({
    model,
    contents,
    config: { temperature: 1.0 },
  });
  return (result.text ?? "").trim();
};

const callGeminiWithRetry = async (
  apiKey: string,
  model: string,
  prompt: { readonly system: string; readonly user: string },
): Promise<string> => {
  for (let attempt = 0; attempt <= GEMINI_MAX_RETRIES; attempt++) {
    try {
      return await callGeminiOnce(apiKey, model, prompt);
    } catch (error) {
      if (attempt < GEMINI_MAX_RETRIES && isRetriableError(error)) {
        const delayMs = GEMINI_BASE_DELAY_MS * 2 ** attempt;
        console.log(`  ⏳ AI fiction retry: model=${model}, attempt=${attempt + 1}, delay=${delayMs}ms`);
        await delay(delayMs);
        continue;
      }
      throw error;
    }
  }
  throw new Error(`Exhausted ${GEMINI_MAX_RETRIES} retries for model ${model}`);
};

export const callGeminiModelChain = async (
  apiKey: string,
  models: readonly string[],
  prompt: { readonly system: string; readonly user: string },
): Promise<{ readonly text: string; readonly model: string }> => {
  for (let i = 0; i < models.length; i++) {
    const model = models[i]!;
    const isLast = i === models.length - 1;
    try {
      const text = await callGeminiWithRetry(apiKey, model, prompt);
      return { text, model };
    } catch (error) {
      console.log(`  ⚠️  Model ${model} failed: ${error instanceof Error ? error.message : String(error)}`);
      if (isLast) throw error;
      console.log(`  🔄 Trying fallback: ${models[i + 1]}`);
    }
  }
  throw new Error("All models exhausted");
};

// ---------------------------------------------------------------------------
// Configuration and result types
// ---------------------------------------------------------------------------

export interface FictionConfig {
  readonly apiKey: string;
  readonly models: readonly string[];
  readonly noteContent: string;
}

export interface FictionResult {
  readonly fiction: string;
  readonly model: string;
  readonly updatedContent: string;
}

// ---------------------------------------------------------------------------
// Main generation function (I/O)
// ---------------------------------------------------------------------------

export const generateFiction = async (
  config: FictionConfig,
): Promise<FictionResult> => {
  const stripped = stripForPrompt(config.noteContent);
  const prompt = buildFictionPrompt(stripped);
  const { text, model } = await callGeminiModelChain(config.apiKey, config.models, prompt);
  const fiction = parseFictionResponse(text);
  const updatedContent = applyFiction(config.noteContent, fiction);

  return { fiction, model, updatedContent };
};

File: haskell/src/Automation/AiFiction.hs
module Automation.AiFiction
  ( fictionSectionHeader
  , defaultFictionModel
  , fictionModelPool
  , fictionGenerationConfig
  , selectFictionModelChain
  , fictionEligibilityCutoff
  , stripForPrompt
  , reflectionNeedsFiction
  , buildFictionPrompt
  , parseFictionResponse
  , buildFictionSignature
  , applyFiction
  , generateFiction
  , FictionConfig (FictionConfig, models, noteContent)
  , FictionResult (FictionResult, fiction, model, updatedContent)
  ) where

import Data.List.NonEmpty (NonEmpty ((:|)))
import qualified Data.List.NonEmpty as NE
import Data.Text (Text)
import qualified Data.Text as T
import Data.Time (Day, LocalTime (..), TimeOfDay (..))
import Data.Time.Calendar (toModifiedJulianDay)

import Automation.Platform
  ( updatesSectionHeader
  )

import Automation.Text (stripCodeFences)
import qualified Automation.Gemini as Gemini
import qualified Automation.Platforms.Bluesky as Bluesky
import qualified Automation.Platforms.Mastodon as Mastodon
import qualified Automation.Platforms.Twitter as Twitter

fictionSectionHeader :: Text
fictionSectionHeader = "## 🤖🐲 AI Fiction"

defaultFictionModel :: Gemini.Model
defaultFictionModel = Gemini.Gemini25Flash

fictionModelPool :: NonEmpty Gemini.Model
fictionModelPool =
  Gemini.Gemini25Flash :|
    [ Gemini.Gemini25FlashLite
    , Gemini.Gemini31FlashLite
    , Gemini.Gemini3Flash
    , Gemini.Gemma4
    , Gemini.Gemma4MixtureOfExperts
    ]

-- | Generation config the daily fiction run sends to Gemini (shared with the
-- daily title generator through @callGeminiForGenerator@). Thinking-capable
-- models (and Gemma, which streams its planning as plain output text) spend a
-- large share of the output-token budget on internal reasoning before emitting
-- the story, so the budget sits well above the under-100-word story length to
-- leave room for the final text. The @test-fiction-models@ binary uses this
-- same config so its output matches what the daily run would actually publish.
fictionGenerationConfig :: Gemini.GenerationConfig
fictionGenerationConfig = Gemini.defaultGenerationConfig
  { Gemini.temperature     = 0.9
  , Gemini.maxOutputTokens = 2048
  }

selectFictionModelChain :: Day -> NonEmpty Gemini.Model -> NonEmpty Gemini.Model
selectFictionModelChain day pool =
  let models = NE.toList pool
      count = length models
      offset = fromInteger (toModifiedJulianDay day `mod` toInteger count)
      rotated = drop offset models <> take offset models
  in case rotated of
    (m : ms) -> m :| ms
    []       -> pool

fictionEligibilityCutoff :: Day -> LocalTime
fictionEligibilityCutoff day = LocalTime day (TimeOfDay 22 0 0)

embedHeaders :: [Text]
embedHeaders = [Twitter.sectionHeader, Bluesky.sectionHeader, Mastodon.sectionHeader]

stripForPrompt :: Text -> Text
stripForPrompt content =
  let contentLines = T.lines content
      hasFrontmatter = case contentLines of
        (first:_) -> T.strip first == "---"
        []        -> False
      fmEnd = if hasFrontmatter
              then findClosingDash (drop 1 contentLines) 1
              else Nothing
      bodyLines = case fmEnd of
        Just index -> drop (index + 1) contentLines
        Nothing  -> contentLines
      body = T.unlines bodyLines
      firstEmbedIdx = findFirstIndex embedHeaders body
      fictionIdx = indexOfHeader fictionSectionHeader body
      updatesIdx = indexOfHeader updatesSectionHeader body
      cutoff = minimum $ filter (>= 0) [firstEmbedIdx, fictionIdx, updatesIdx, T.length body]
  in T.strip $ T.take cutoff body

findClosingDash :: [Text] -> Int -> Maybe Int
findClosingDash [] _ = Nothing
findClosingDash (line:rest) index =
  if T.strip line == "---" then Just index else findClosingDash rest (index + 1)

findFirstIndex :: [Text] -> Text -> Int
findFirstIndex headers body =
  let indices = fmap (`indexOfHeader` body) headers
      valid = filter (>= 0) indices
  in case valid of
    [] -> T.length body
    _  -> minimum valid

indexOfHeader :: Text -> Text -> Int
indexOfHeader header body =
  case T.breakOn header body of
    (before, rest) | not (T.null rest) -> T.length before
    _ -> -1

reflectionNeedsFiction :: Text -> Bool
reflectionNeedsFiction content = not (T.isInfixOf fictionSectionHeader content)

buildFictionPrompt :: Text -> (Text, Text)
buildFictionPrompt strippedContent =
  let system = "You are an inventive fiction writer crafting one tiny story per day.\n\
               \\n\
               \Your task:\n\
               \1. Read today's daily reflection note below\n\
               \2. Find the SINGLE most distinctive thread in today's topics — the one image, tension, question, or detail that makes TODAY different from every other day\n\
               \3. Write a short, vivid piece of fiction (UNDER 100 words) that grows out of that specific thread\n\
               \\n\
               \VOICE AND VIBE:\n\
               \- Write an engaging FIRST-PERSON narrative, or a tightly-scoped THIRD-PERSON narrative that stays close to a single character\n\
               \- Ground it in a concrete scene: a real place, a specific moment, a character actually doing something\n\
               \- Let the theme fall out of the specifics of today's topics so that each day feels unmistakably its own\n\
               \- Favor sensory detail, a small narrative arc, and a distinct narrator over abstract, generic musing\n\
               \- AVOID stale, interchangeable imagery (shifting sands, humming networks, woven tapestries, whispering winds, grand looms, glittering surfaces). Surprise the reader instead.\n\
               \\n\
               \FORMATTING RULES:\n\
               \- Every sentence MUST begin with an emoji that fits that sentence\n\
               \- NEVER use quotation marks (no \", no ', no `, no curly quotes)\n\
               \- Keep it UNDER 100 words\n\
               \- Do NOT name or directly quote the specific books, videos, articles, or people — transform them into the story\n\
               \- Do NOT include any heading or title — output ONLY the fiction text\n\
               \- Output ONLY the fiction — no explanation, no theme list, no preamble"
      user = "Write today's short fiction, inspired by the most distinctive thread in this daily reflection:\n\n" <> strippedContent
  in (system, user)

parseFictionResponse :: Text -> Text
parseFictionResponse raw =
  let t1 = stripCodeFences raw
      t2 = T.replace fictionSectionHeader "" t1
      t3 = removeQuotationMarks t2
  in T.strip t3

removeQuotationMarks :: Text -> Text
removeQuotationMarks = T.filter (\character ->
  character /= '"' && character /= '\'' && character /= '`'
  && character /= '\x2018' && character /= '\x2019'
  && character /= '\x201c' && character /= '\x201d')

buildFictionSignature :: Text -> Text
buildFictionSignature model = "\n\n✍️ Written by " <> model

applyFiction :: Text -> Text -> Maybe Text -> Text
applyFiction content fiction mModel =
  let signature = maybe "" buildFictionSignature mModel
      sectionBlock = fictionSectionHeader <> "\n\n" <> fiction <> signature
      embedIndices = filter (>= 0) $ fmap (`indexOfHeader` content) embedHeaders
      updatesIdx = indexOfHeader updatesSectionHeader content
      candidates = filter (>= 0) (updatesIdx : embedIndices)
      insertBeforeIdx = case candidates of
        [] -> -1
        _  -> minimum candidates
  in if insertBeforeIdx >= 0
    then
      let before = T.stripEnd (T.take insertBeforeIdx content)
          after = T.drop insertBeforeIdx content
      in before <> "\n\n" <> sectionBlock <> "\n\n" <> after
    else
      T.stripEnd content <> "\n\n" <> sectionBlock <> "\n"

data FictionConfig = FictionConfig
  { models      :: NonEmpty Gemini.Model
  , noteContent :: Text
  } deriving (Show, Eq)

data FictionResult = FictionResult
  { fiction        :: Text
  , model          :: Text
  , updatedContent :: Text
  } deriving (Show, Eq)

generateFiction :: FictionConfig -> (NonEmpty Gemini.Model -> (Text, Text) -> IO (Text, Text)) -> IO FictionResult
generateFiction config callModel = do
  let stripped = stripForPrompt (noteContent config)
      prompt = buildFictionPrompt stripped
  (text, usedModel) <- callModel (models config) prompt
  let fictionText = parseFictionResponse text
      newContent = applyFiction (noteContent config) fictionText (Just usedModel)
  pure FictionResult
    { fiction = fictionText
    , model = usedModel
    , updatedContent = newContent
    }
